From 0c06ca296791538459c589f0983c48e284aea612 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bernhard=20=C3=9Cbelacker?= <bernhardu@mailbox.org>
Date: Tue, 17 Dec 2024 23:47:37 +0100
Subject: [PATCH 49/86] shell32: Use no_sanitize_address in ShellView_WndProc.

/home/bernhard/data/entwicklung/2024/wine/wine/tools/runtest -q -P wine -T . -M shell32.dll -p dlls/shell32/tests/i386-windows/shell32_test.exe shlview && touch dlls/shell32/tests/i386-windows/shlview.ok
=================================================================
==1076==ERROR: AddressSanitizer: heap-use-after-free on address 0x02b03524 at pc 0x775fbc42 bp 0x00b5e894 sp 0x00b5e890
READ of size 4 at 0x02b03524 thread T0
0404:fixme:msvcrt:_set_abort_behavior _WRITE_CALL_REPORTFAULT unhandled
0404:fixme:file:server_get_file_info Unsupported info class e
    #0 0x775fbc41 in ShellView_WndProc /home/bernhard/data/entwicklung/2024/wine\wine/dlls/shell32/shlview.c:1700:37
    #1 0x75aa8787 in WINPROC_wrapper (C:\windows\system32\user32.dll+0x10008787)
    #2 0x75afb658 in call_window_proc /home/bernhard/data/entwicklung/2024/wine/wine/dlls/user32\winproc.c:111:15
    #3 0x75b00f90 in dispatch_win_proc_params /home/bernhard/data/entwicklung/2024/wine/wine/dlls/user32\winproc.c:710:13
    #4 0x75aedf7a in User32CallWindowProc /home/bernhard/data/entwicklung/2024/wine/wine/dlls/user32\winproc.c:824:14
    #5 0x789328f8 in dispatch_user_callback /home/bernhard/data/entwicklung/2024/wine/wine/dlls/ntdll\exception.c:292:18
    #6 0x78900b6e in KiUserCallbackDispatcher /home/bernhard/data/entwicklung/2024/wine/wine/dlls/ntdll\signal_i386.c:222:23
    #7 0x757adf97 in NtUserDestroyWindow (C:\windows\system32\win32u.dll+0x1000df97)

0x02b03524 is located 100 bytes inside of 148-byte region [0x02b034c0,0x02b03554)
freed by thread T0 here:
    #0 0x7645281b in free /home/bernhard/data/entwicklung/2024/llvm-mingw/2024-10-18/llvm-mingw/llvm-project/compiler-rt\lib/asan/asan_malloc_win.cpp:71:3
    #1 0x775f38d3 in IShellView_fnRelease /home/bernhard/data/entwicklung/2024/wine\wine/dlls/shell32/shlview.c:1812:4
    #2 0x775f8e94 in ShellView_WndProc /home/bernhard/data/entwicklung/2024/wine\wine/dlls/shell32/shlview.c:1699:7
    #3 0x75aa8787 in WINPROC_wrapper (C:\windows\system32\user32.dll+0x10008787)
    #4 0x75afb658 in call_window_proc /home/bernhard/data/entwicklung/2024/wine/wine/dlls/user32\winproc.c:111:15
    #5 0x75b00f90 in dispatch_win_proc_params /home/bernhard/data/entwicklung/2024/wine/wine/dlls/user32\winproc.c:710:13
    #6 0x75aedf7a in User32CallWindowProc /home/bernhard/data/entwicklung/2024/wine/wine/dlls/user32\winproc.c:824:14
    #7 0x789328f8 in dispatch_user_callback /home/bernhard/data/entwicklung/2024/wine/wine/dlls/ntdll\exception.c:292:18
    #8 0x78900b6e in KiUserCallbackDispatcher /home/bernhard/data/entwicklung/2024/wine/wine/dlls/ntdll\signal_i386.c:222:23
    #9 0x757adf97 in NtUserDestroyWindow (C:\windows\system32\win32u.dll+0x1000df97)

previously allocated by thread T0 here:
    #0 0x76452a1e in calloc /home/bernhard/data/entwicklung/2024/llvm-mingw/2024-10-18/llvm-mingw/llvm-project/compiler-rt\lib/asan/asan_malloc_win.cpp:91:3
    #1 0x775f1bb3 in IShellView_Constructor /home/bernhard/data/entwicklung/2024/wine\wine/dlls/shell32/shlview.c:3789:10

SUMMARY: AddressSanitizer: heap-use-after-free /home/bernhard/data/entwicklung/2024/wine\wine/dlls/shell32/shlview.c:1700:37 in ShellView_WndProc
Shadow bytes around the buggy address:
  0x02b03280: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x02b03300: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x02b03380: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x02b03400: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x02b03480: fa fa fa fa fa fa fa fa fd fd fd fd fd fd fd fd
=>0x02b03500: fd fd fd fd[fd]fd fd fd fd fd fd fa fa fa fa fa
  0x02b03580: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fd
  0x02b03600: fd fd fd fd fd fd fd fa fa fa fa fa fa fa fa fa
  0x02b03680: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x02b03700: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x02b03780: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==1076==ABORTING
make: *** [Makefile:370069: dlls/shell32/tests/i386-windows/shlview.ok] Fehler 1
---
 dlls/shell32/shlview.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/shell32/shlview.c b/dlls/shell32/shlview.c
index 9c699ecda08..6b26520de76 100644
--- a/dlls/shell32/shlview.c
+++ b/dlls/shell32/shlview.c
@@ -1657,6 +1657,7 @@ static LRESULT ShellView_OnChange(IShellViewImpl * This, const LPCITEMIDLIST *pi
 *  ShellView_WndProc
 */
 
+__attribute__((no_sanitize_address))
 static LRESULT CALLBACK ShellView_WndProc(HWND hWnd, UINT uMessage, WPARAM wParam, LPARAM lParam)
 {
 	IShellViewImpl * pThis = (IShellViewImpl*)GetWindowLongPtrW(hWnd, GWLP_USERDATA);
-- 
2.47.1

